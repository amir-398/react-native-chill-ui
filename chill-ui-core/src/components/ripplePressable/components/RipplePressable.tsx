import { RipplePressablePropsTw } from '@types';
import { AnimatedBox } from '@components/animatedBox';
import { View, Animated, Pressable } from 'react-native';
import { cn, classNamePropsHandler, classNameHandler, styleHandler } from '@utils';
import { useState, useRef, useEffect, PropsWithChildren, forwardRef } from 'react';

import styles from '../styles/RipplePressable.ss.styles';
import { extractBorderRadius } from '../utils/extractBorder';
import { twStyles } from '../styles/RipplePressable.tw.styles';

/**
 * Individual ripple effect component using native Animated API
 */
interface RippleEffectProps {
  x: number;
  y: number;
  duration: number;
  effectColor: string;
  containerWidth: number;
  containerHeight: number;
}

function RippleEffect({ containerHeight, containerWidth, duration, effectColor, x, y }: RippleEffectProps) {
  const scaleAnim = useRef(new Animated.Value(0)).current;
  const opacityAnim = useRef(new Animated.Value(0.6)).current;

  const distanceToCorners = [
    Math.sqrt(x * x + y * y), // Top-left
    Math.sqrt((containerWidth - x) * (containerWidth - x) + y * y), // Top-right
    Math.sqrt(x * x + (containerHeight - y) * (containerHeight - y)), // Bottom-left
    Math.sqrt((containerWidth - x) * (containerWidth - x) + (containerHeight - y) * (containerHeight - y)), // Bottom-right
  ];

  const maxDistance = Math.max(...distanceToCorners);

  const baseSize = 20;
  const finalScale = (maxDistance * 2.2) / baseSize;

  useEffect(() => {
    Animated.parallel([
      Animated.timing(scaleAnim, {
        duration,
        toValue: finalScale,
        useNativeDriver: true,
      }),
      Animated.timing(opacityAnim, {
        delay: Math.round(duration * 0.3),
        duration: Math.round(duration * 0.9),
        toValue: 0,
        useNativeDriver: true,
      }),
    ]).start();
  }, [scaleAnim, opacityAnim, finalScale, duration]);

  return (
    <AnimatedBox
      style={{
        backgroundColor: effectColor,
        borderRadius: baseSize / 2,
        height: baseSize,
        left: x - baseSize / 2,
        opacity: opacityAnim,
        pointerEvents: 'none',
        position: 'absolute',
        top: y - baseSize / 2,
        transform: [{ scale: scaleAnim }],
        width: baseSize,
      }}
    />
  );
}

/**
 * The `<RipplePressable />` component provides a ripple effect on press.
 * Creates a visual ripple animation that emanates from the touch point.
 *
 *
 * <!-- STORYBOOK_IMPORT_START
 * ```tsx
 * import { RipplePressable } from 'react-native-chill-ui';
 * ```
 * STORYBOOK_IMPORT_END -->
 *
 * @example
 * ```tsx
 * <RipplePressable onPress={() => console.log('Pressed!')}>
 *   <Box className="p-4 bg-blue-500 rounded-lg">
 *     <String color="white">Press me</String>
 *   </Box>
 * </RipplePressable>
 * ```
 *
 * @param children - Child components to render with ripple effect (required)
 * @param className - Custom CSS classes for styling the container (NativeWind)
 * @param disabled - Whether the component is disabled (default: false)
 * @param effectColor - Color of the ripple effect (default: 'rgba(255, 255, 255, 0.6)')
 * @param onPress - Callback function called when the component is pressed, receives press event
 * @param speed - Animation duration in milliseconds (default: 500)
 * @param style - Style object for the pressable container
 */
const RipplePressable = forwardRef<View, PropsWithChildren<RipplePressablePropsTw>>((props, ref) => {
  classNamePropsHandler(props, 'RipplePressable');
  const {
    children,
    className,
    disabled = false,
    effectColor = 'rgba(255, 255, 255, 0.6)',
    onPress,
    speed = 500,
    style,
    ...rest
  } = props;

  if (!children) {
    throw new Error('RipplePressable must have children');
  }

  const [ripples, setRipples] = useState<
    { id: number; x: number; y: number; timestamp: number; width: number; height: number }[]
  >([]);
  const containerRef = useRef<View>(null);
  const rippleIdCounter = useRef(0);

  // Extract borderRadius from children
  const childBorderRadius = extractBorderRadius(children);

  /**
   * Handles press event and creates ripple animation
   */
  const handlePress = (event: any) => {
    if (disabled) {
      return;
    }

    if (containerRef.current) {
      containerRef.current.measure((_x, _y, width, height) => {
        const touchX = event.nativeEvent.locationX;
        const touchY = event.nativeEvent.locationY;

        rippleIdCounter.current += 1;
        const rippleId = rippleIdCounter.current;
        const newRipple = {
          height,
          id: rippleId,
          timestamp: Date.now(),
          width,
          x: touchX,
          y: touchY,
        };

        setRipples(prevRipples => [...prevRipples, newRipple]);

        setTimeout(() => {
          setRipples(prevRipples => prevRipples.filter(ripple => ripple.id !== rippleId));
        }, speed + 150);
      });
    }

    if (onPress) {
      onPress(event);
    }
  };

  return (
    <Pressable
      {...rest}
      ref={ref || containerRef}
      onPress={handlePress}
      disabled={disabled}
      {...classNameHandler(cn(twStyles.container, disabled && twStyles.disabled, className))}
      {...styleHandler({
        defaultStyle: [styles.container, { borderRadius: childBorderRadius }, disabled && styles.disabled],
        style,
      })}
    >
      {children}
      {!disabled &&
        ripples.map(ripple => (
          <RippleEffect
            key={ripple.id}
            x={ripple.x}
            y={ripple.y}
            containerWidth={ripple.width}
            containerHeight={ripple.height}
            duration={speed}
            effectColor={effectColor}
          />
        ))}
    </Pressable>
  );
});

RipplePressable.displayName = 'RipplePressable';

export default RipplePressable;
