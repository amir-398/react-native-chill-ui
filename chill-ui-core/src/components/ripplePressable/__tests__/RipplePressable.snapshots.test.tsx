import { render } from '@testing-library/react-native';
import { Text, View } from 'react-native';
import RipplePressable from '../components/RipplePressable';

// Mock utils
jest.mock('../../../utils', () => ({
  classNameHandler: jest.fn(() => ({})),
  classNamePropsHandler: jest.fn(),
  cn: jest.fn((...args) => args.filter(Boolean).join(' ')),
  colorVariantPropsHandler: jest.fn(),
  isString: jest.fn(value => typeof value === 'string'),
  styleHandler: jest.fn(() => ({})),
}));

// Mock AnimatedBox
jest.mock('../../../components/animatedBox', () => ({
  AnimatedBox: ({ children, ...props }: any) => {
    const React = require('react');
    const { View } = require('react-native');
    return (
      <View testID="animated-box" {...props}>
        {children}
      </View>
    );
  },
}));

describe('RipplePressable Snapshots', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot with basic props', () => {
    const tree = render(
      <RipplePressable>
        <Text>Basic Content</Text>
      </RipplePressable>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with custom effectColor', () => {
    const tree = render(
      <RipplePressable effectColor="rgba(255, 0, 0, 0.3)">
        <Text>Red Ripple</Text>
      </RipplePressable>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with custom speed', () => {
    const tree = render(
      <RipplePressable speed={300}>
        <Text>Fast Ripple</Text>
      </RipplePressable>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with disabled state', () => {
    const tree = render(
      <RipplePressable disabled>
        <Text>Disabled</Text>
      </RipplePressable>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with custom styling', () => {
    const tree = render(
      <RipplePressable className="custom-class" style={{ margin: 10 }}>
        <View style={{ padding: 16, backgroundColor: '#3B82F6', borderRadius: 8 }}>
          <Text style={{ color: 'white' }}>Styled Button</Text>
        </View>
      </RipplePressable>,
    );
    expect(tree).toMatchSnapshot();
  });
});
