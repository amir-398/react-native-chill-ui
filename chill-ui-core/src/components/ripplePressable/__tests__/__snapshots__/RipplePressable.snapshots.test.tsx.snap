// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RipplePressable Snapshots should match snapshot with basic props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text>
    Basic Content
  </Text>
</View>
`;

exports[`RipplePressable Snapshots should match snapshot with custom effectColor 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text>
    Red Ripple
  </Text>
</View>
`;

exports[`RipplePressable Snapshots should match snapshot with custom speed 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text>
    Fast Ripple
  </Text>
</View>
`;

exports[`RipplePressable Snapshots should match snapshot with custom styling 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      {
        "backgroundColor": "#3B82F6",
        "borderRadius": 8,
        "padding": 16,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
        }
      }
    >
      Styled Button
    </Text>
  </View>
</View>
`;

exports[`RipplePressable Snapshots should match snapshot with disabled state 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text>
    Disabled
  </Text>
</View>
`;
